src/validation/rules/optional/ProhibitDeprecatedFieldsRule.d.ts

```diff
@@ -0,0 +1,15 @@
+import { ASTVisitor } from '../../../language/visitor';
+import { ValidationContext } from '../../ValidationContext';
+
+/**
+ * Prohibit deprecated fields
+ *
+ * A GraphQL document is only valid if all selected fields and all used enum
+ * values have not been deprecated.
+ *
+ * Note: This rule is optional and is not part of the Validation section of the
+ * GraphQL Specification.
+ */
+export function ProhibitDeprecatedFieldsRule(
```

Comment:
I wouldn't use "fields" since we also forbid enum values and plan to support deprecation on arguments so we need a more generic name.

Reply:
Hmm. Is there a name for "things in the schema"? I think the closest I've seen in the spec is "things" and "portions of the schema".

`ProhibitDeprecatedUsages`?

Reply:
`ProhibitDeprecatedUsage` sounds reasonable.

@benjie You are really good at coming up with good names. Maybe you can advise?
We creating a folder with optional rules and two first rules would be `ProhibitIntrospectionQueriesRule` and `ProhibitDeprecatedUsages`.
Does this naming scheme make sense for these two rules and similar rules in the future?

Reply:
All the current rules describe the desired outcome, what the valid query has, rather than the action of the rule itself. And they end in “Rule”. So in this we want a valid query to have no deprecated values, fields, etc... unless we can think of a good collective noun; I think calling it `NoDeprecatedRule` would be acceptable. I’d rather add a word between deprecated and rule, but I can’t think of a good one: usages, things, entities, constructs, fields, components, productions; none of these are quite right. Maybe “elements”?

Reply:
I’d go with `NoIntrospectionRule` for the other; but presumably it doesn’t block __typename.

Reply:
@benje Thanks for review 👍 

> I’d go with NoIntrospectionRule for the other; but presumably it doesn’t block __typename.

Yes, that's why we added `Queries`.

>  I think calling it NoDeprecatedRule would be acceptable.

I kind of wanted to separate it from other rules since other `No*` mean it's forbidden by the spec idea was to show that you forbid something that is legal but it's just your decision to do so.
Do you think it's worth having this separation?

Reply:
> Yes, that's why we added Queries.

I don't feel adding `Queries` really adds any difference between `__typename` vs `__type` here. E.g. `{ currentUser { __typename id friends { name } teams { name } } __type(name: "UserType") { enumValues { name } } }` is not an "introspection query" exactly, but it does contain introspection. I think `NoIntrospectionRule` is simpler and more direct, and it's fairly obvious to most GraphQL users that `__typename` (though an introspection field technically) would not be forbidden.

If you want a more precise name, `NoSchemaIntrospectionRule` would be a better choice, IMO.

> I kind of wanted to separate it from other rules since other No* mean it's forbidden by the spec idea was to show that you forbid something that is legal but it's just your decision to do so.

I don't think having a subtle language difference like this is a good way of demarking this separation. I'd stick to the existing naming conventions; but potentially prefix the rule with `Custom` or similar text*, i.e. `CustomNoSchemaIntrospectionRule`.

Strictly the `NoSchemaIntrospectionRule` actually breaks the GraphQL spec (which requires introspection be available), so I'm not really happy with something like "Optional" as a description for it, hence `Custom`.

* `Custom`: `NonStandard`, `OptIn`, `Optional`, `Bespoke`, `Unspecified`, `Extra`, `Unsanctioned`, ...

> Do you think it's worth having this separation?

It seems like a good idea if it can be executed well.

Reply:
@benjie Thanks `NoSchemaIntrospectionRule` looks great 👍 

>  so I'm not really happy with something like "Optional" as a description for it, hence Custom.

Great logic 👍  Actually refusing deprecated rules is also breaking the spec.
How about moving `Custom` to the end so `NoSchemaIntrospectionCustomRule` and `NoDeprecatedCustomRule`?

Reply:
Sure, sounds good :+1:

Reply:
@danielrearden What do you think about the proposed naming schema?
So `NoSchemaIntrospectionCustomRule` and `NoDeprecatedCustomRule` will live in `src/validation/rules/custom`.

Reply:
@IvanGoncharov Sounds great. I'll get both PRs updated and address the rest of the comments in this one ASAP. Thanks for the help @benjie !

Reply:
@danielrearden Don't hurry we still blocked by TS conversion of tests. Don't want to prolong it so we in freeze for everything unrelated that doesn't touch tests probably for next couple days.

Reply:
I figured, also why I'm holding off on breaking up the harness schema like we discussed

Reply:
@danielrearden If you have time for this it would be great if you can convert tests for `express-graphql` into TS meanwhile. should be way simpler but try to separate changes into PRs (.d.ts updates (if needed), missing types (you can start without `noImplicitAny` and enable it in the next PR) so basically try to separate syntax changes from something serious.